---
title: "Single Cell DIPG"
author: "Maximilian Lombardo"
date: "10/15/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}

#Two different single glioma data sets from regev
#GSE57872_GBM_data_matrix.txt is from the following paper: Single cell RNA-seq of primary human glioblastomas, GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE57872
#GSE102130_K27Mproject.RSEM.vh20170621.txt is from the following paper: Single cell RNA-seq analysis of K27M-mutant glioma, GEO: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE102130


sc.data <- readRegevData(file.path = "~/Documents/uva/master/data/RegevSingleCell/GSE102130_K27Mproject.RSEM.vh20170621.txt")

```


```{r}

library(Seurat)

# Initialize the Seurat object with the raw (non-normalized data).  Keep all
# genes expressed in >= 3 cells (~0.1% of the data). Keep all cells with at
# least 200 detected genes
dipg <- CreateSeuratObject(raw.data = sc.data, min.cells = 3, min.genes = 200, project = "HiSeq500")



# The number of genes and UMIs (nGene and nUMI) are automatically calculated
# for every object by Seurat.  For non-UMI data, nUMI represents the sum of
# the non-normalized values within a cell We calculate the percentage of
# mitochondrial genes here and store it in percent.mito using AddMetaData.
# We use object@raw.data since this represents non-transformed and
# non-log-normalized counts The % of UMI mapping to MT-genes is a common
# scRNA-seq QC metric.
mito.genes <- grep(pattern = "^MT-", x = rownames(x = dipg@data), value = TRUE)#no mito genes????
percent.mito <- Matrix::colSums(dipg@raw.data[mito.genes, ])/Matrix::colSums(dipg@raw.data)

# AddMetaData adds columns to object@meta.data, and is a great place to
# stash QC stats
dipg <- AddMetaData(object = dipg, metadata = percent.mito, col.name = "percent.mito")
VlnPlot(object = dipg, features.plot = c("nGene", "nUMI", "percent.mito"), nCol = 3)

#Visualize Gene-Gene Relationships
par(mfrow = c(1, 2))
GenePlot(object = dipg, gene1 = "nUMI", gene2 = "percent.mito")
GenePlot(object = dipg, gene1 = "nUMI", gene2 = "nGene")


#Normalize the data 

dipg <- NormalizeData(object = dipg, normalization.method = "LogNormalize", scale.factor = 10000)

dipg <- FindVariableGenes(object = dipg, mean.function = ExpMean, dispersion.function = LogVMR, 
    x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5)

#Scale the data
dipg <- ScaleData(object = dipg, vars.to.regress = c("nUMI"))



dipg <- RunPCA(object = dipg, pc.genes = dipg@var.genes, pcs.compute = 100, do.print = FALSE)


PCElbowPlot(object = dipg, num.pc = 100)

##PCS to use == 25 for diffuse midline glioma data set????

#TSNE Plot
dipg <- RunTSNE(object = dipg, dims.use = 1:25, do.fast = TRUE)

dipg <- FindClusters(object = dipg, reduction.type = "pca", dims.use = 1:25, 
    resolution = 0.6, print.output = 0, save.SNN = TRUE)


```


```{r}
dipg@meta.data$Batch <- unlist(lapply(dipg@cell.names, FUN = function(cell.name){unlist(strsplit(cell.name, split = "\\."))[1]}))
```
